### PR TITLE
Label the projects created by cluster-loader

### PR DESCRIPTION
**Pull request by hongkailiu**
This feature would be useful for manual clean-up, such as `oc delete project -l test=purpose` if we have the labels defined in the config file:

```
provider: local
ClusterLoader:
  cleanup: false
  projects:
    - num: 2
      basename: clusterproject
      labels:
         purpose: test
```

### CODE REVIEW

**Comment by mffiedler** [comment]
@sjug PTAL

File: test/extended/cluster/cl.go [hunk 1]
```diff
@@ -202,6 +208,19 @@ func newProject(nsName string) *projectapi.Project {
 	}
 }
 
+func labelNamespace(c kclientset.Interface, ns string, pairs map[string]string) (*v1.Namespace, error) {
```
Comment (sjug) [inline test/extended/cluster/cl.go]: Please move this into `utils.go`, you can see there are a host of dependencies being imported here (in this commit) that for the most part already are in utils.

The function name is unclear to me... Are we trying to `SetNamespaceLabels` or `UpdateNamespaceLabels`?

File: test/extended/cluster/cl.go [hunk 2]
```diff
@@ -202,6 +208,19 @@ func newProject(nsName string) *projectapi.Project {
...
+	if len(pairs) == 0 {
+		return nil, nil
+	}
+	var d struct{ Metadata metav1.ObjectMeta `json:"metadata"`}
```
Comment (sjug) [inline test/extended/cluster/cl.go]: I've never seen a nested struct created like this, but I think the underlying issue is that you're trying to take existing golang types and then convert them to json which is unnecessary. The `Namespace.Labels` are already of type `map[string]string` which is the same type you've added to the `ClusterLoaderType`.

File: test/extended/cluster/cl.go [hunk 3]
```diff
@@ -202,6 +208,19 @@ func newProject(nsName string) *projectapi.Project {
...
+	d.Metadata = metav1.ObjectMeta{Labels:pairs}
+	bytes,err := json.Marshal(d)
```
Comment (sjug) [inline test/extended/cluster/cl.go]: This is totally unecessary.

File: test/extended/cluster/cl.go [hunk 4]
```diff
@@ -202,6 +208,19 @@ func newProject(nsName string) *projectapi.Project {
...
+	if err != nil {
+		return nil, err
+	}
+	return c.CoreV1().Namespaces().Patch(ns, types.MergePatchType, bytes)
```
Comment (sjug) [inline test/extended/cluster/cl.go]: An easy way to do this with native objects would be to get the existing namespace pointer with:
```
CoreV1().Namespaces().Get(ns, metav1.GetOptions{})
```
Modify the `Namespace.Label` directly by appending or setting the value to `pairs`.

Then just update the namespace:
```
CoreV1().Namespaces().Update(Namespace)
```
where `Namespace` is the modified object you recieved from the first `Get` and which you updated with `pairs`.

**Comment by hongkailiu** [comment]
@sjug 
In a training this week.
Will update early next week.

File: test/extended/cluster/cl.go (hunk 1 above)
Comment (hongkailiu) [reply test/extended/cluster/cl.go]: Done.
Moved and renamed to SetNamespaceLabels

File: test/extended/cluster/cl.go (hunk 4 above)
Comment (hongkailiu) [reply test/extended/cluster/cl.go]: Done.
With a little concern/question:
Update method uses the whole ns object as the posting body while patch method uses only the label/metadata part. Probably the latter is a bit more efficient.

I tried "oc label project ... --loglelvel=10" and saw only metadata is in the body of http-req.
That was why I used patch method in the first commit here.

Maybe namespaces/projects are not that big, specially for the ones we have just created.
We do not need to worry about that.

**Comment by hongkailiu** [comment]
Done.
Moved and renamed to SetNamespaceLabels

With a little concern/question:
Update method uses the whole ns object as the posting body while patch method uses only the label/metadata part. Probably the latter is a bit more efficient.

I tried "oc label project ... --loglelvel=10" and saw only metadata is in the body of http-req.
That was why I used patch method in the first commit here.

Maybe namespaces/projects are not that big, specially for the ones we have just created.
We do not need to worry about that.

**Comment by sjug** [comment]
@hongkailiu 
I'm not sure how you define efficient...

In this solution `Get()` call returns a pointer not a new object. Then you're updating a field within the struct in that pointer, and calling an `Update()`. 

In your version you were creating new objects and then marshalling them to json, before making the `Patch()` call which will then need to unmarshall the json on the other end.

File: test/extended/cluster/cl.go [hunk 5]
```diff
@@ -79,6 +79,9 @@ var _ = g.Describe("[Feature:Performance][Serial][Slow] Load cluster", func() {
 				err := oc.Run("new-project").Args(nsName).Execute()
 				o.Expect(err).NotTo(o.HaveOccurred())
 				e2e.Logf("%d/%d : Created new namespace: %v", j+1, p.Number, nsName)
+				// label namespace nsName
+				_, err = SetNamespaceLabels(c, nsName, p.Labels)
```
Comment (sjug) [inline test/extended/cluster/cl.go]: If you check the other `Create{Configmaps,Secrets}` calls we like to check if the object exists before making the function call. I see that you have added the check in the `SetNamespaceLabels`, but could we (also?) have the check in line here (as below)?

**Comment by hongkailiu** [comment]
Done.
Will clarify what I think about the patch and update offline.

**Comment by sjug** [comment]
Hey @hongkailiu 

```
[WARNING] !!! gofmt needs to be run on the listed files
[FATAL] Try running 'gofmt -s -d [path]'
[FATAL] Or autocorrect with 'hack/verify-gofmt.sh | xargs -n 1 gofmt -s -w'
[ERROR] hack/verify-gofmt.sh exited with code 1 after 00h 00m 06s
```

**Comment by hongkailiu** [comment]
@smarterclayton @bparees
Let me know what I need to do before merging?
Thanks.

**Pull request merged**